Extract entity ID when creating SAML service provider.

mdwn opened the pull request:
The entity ID will be parsed from the entity descriptor metadata when the service provider's entity descriptor is set. The motivation of this PR is to do the following:

* Extract the entity ID from the entity descriptor during the creation of the SAML IdP service provider.
* Prevent the creation of service providers in the backend that have an entity ID that does not match the entity ID represented in the entity descriptor.
* When reading or listing service providers, prevent the parsing of every entity descriptor during the listing.

The goal here is to make it performant to look up entity descriptors by their entity ID without incurring the cost of parsing every single entity descriptor to determine the entity ID.

Commit:
Extract entity ID when creating SAML service provider.

The entity ID will be parsed from the entity descriptor metadata when the
service provider's entity descriptor is set. The motivation of this PR is to
do the following:

* Extract the entity ID from the entity descriptor during the creation of the
  SAML IdP service provider.
* Prevent the creation of service providers in the backend that have an
  entity ID that does not match the entity ID represented in the entity
  descriptor.
* When reading or listing service providers, prevent the parsing of every
  entity descriptor during the listing.

The goal here is to make it performant to look up entity descriptors by their
entity ID without incurring the cost of parsing every single entity descriptor
to determine the entity ID.

espadolini commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -94,9 +94,41 @@ func (s *SAMLIdPServiceProviderService) GetSAMLIdPServiceProvider(ctx context.Co
+		return trace.BadParameter("entity ID is set to %s but the entity ID extracted from the provider is %s", currentEntityID, sp.GetEntityID())
+	}
+
+	// Make sure no other service provider has the same entity ID.
```
Is this a lint, or an actual invariant that we must maintain for security reasons?

mdwn replied:
This is a lint.

mdwn replied:
I think I'll have the key be the entity ID and try to derive it during creation from the command line. Thoughts on this?

One thing is that the entity ID is not valid as a `name` in Teleport as it's a URI.

mdwn replied:
So this doesn't really apply any more, and the discussion has been hashed out on Slack a good bit -- the key has remained the same and I've stripped down the extraction code to only really apply during the Create call.

mdwn commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -4873,6 +4873,8 @@ message SAMLIdPServiceProviderV1 {
 message SAMLIdPServiceProviderSpecV1 {
   // EntityDescriptor is the entity descriptor for the service provider
   string EntityDescriptor = 1 [(gogoproto.jsontag) = "entity_descriptor"];
+  // EntityID is the entity ID for the entity descriptor. Please do not set this directly
```
It feels weird to make this required when it's explicitly derived/derivable from the entity descriptor itself. I'm thinking that it makes more sense to have this as the key as you suggested, which we can change (this hasn't been shipped anywhere yet), but the key is a URI. Is that copacetic in our backend?

mdwn replied:
I've changed this a good bit, with the idea that the create validates that the entity ID doesn't already exist in the backend.

Commit:
Feedback, eliminating the parsing from the object, pushing parsing verification into the local service.

mdwn commented:
So I've removed the parsing from the object itself, changed the semantics on the parsing verification in the backend, and added a lock into creation. Let me know what you think of this -- it's difficult to think of how to maintain uniqueness across two fields (both name and entity ID) with our current backend setup.

mdwn committed:
Merge branch 'master' into mike.wilson/store-entity-id-in-saml-id-provider

Commit:
Fix up tests, e_imports.

Commit:
Fix comment in types.proto.

Commit:
Remove dead comment.

Commit:
Deduplicate entity ID in update as well.

Commit:
Small refactor to move parse out of lock.

Commit:
Remove XML parsing from type, fix grpcserver test, fix locking boundaries.

espadolini commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -18,15 +18,21 @@ package local
 )
 
-const samlIDPServiceProviderMaxPageSize = 200
+const (
+	samlIDPServiceProviderModifyLock    = "saml_idp_service_provider_modify_lock"
+	samlIDPServiceProviderModifyLockTTL = time.Second * 5
+	samlIDPServiceProviderMaxPageSize   = 200
+)
```
I'd move this together with `samlIDPServiceProviderPrefix` (or move that one up here).

espadolini commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -97,43 +103,59 @@ func (s *SAMLIdPServiceProviderService) CreateSAMLIdPServiceProvider(ctx context
+			return trace.Wrap(err)
+		}
+
+		value, err := services.MarshalSAMLIdPServiceProvider(sp)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		item := backend.Item{
+			Key:     backend.Key(samlIDPServiceProviderPrefix, sp.GetName()),
+			Value:   value,
+			Expires: sp.Expiry(),
+			ID:      sp.GetResourceID(),
+		}
```
I'd prepare the `item` outside of the locked section, just as a matter of cleanliness; same applies below.

mdwn commented on api/types/saml_idp_service_provider.go:
```diff
@@ -35,6 +31,10 @@ type SAMLIdPServiceProvider interface {
+	// GetEntityID returns the entity ID.
+	GetEntityID() string
+	// SetEntityID sets the entity ID.
+	SetEntityID(string)
```
Yes, this will be used by the SAML IdP (enterprise feature).

It should always be pulled from the descriptor, however in order to look it up we need to either have it pre-parsed (like this PR) or parse every single descriptor to find the entity ID when we're looking for it. This will save us from needing to do that parse every time.

mdwn commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -161,3 +183,46 @@ func (s *SAMLIdPServiceProviderService) DeleteAllSAMLIdPServiceProviders(ctx con
+			// Only check entity ID duplicates if we're looking at objects other than the one we're trying to validate.
+			// This ensures updates will work and that creates will return an already exists error.
+			if listSp.GetName() != sp.GetName() && listSp.GetEntityID() == sp.GetEntityID() {
+				return trace.BadParameter("service provider %s has the same entity ID %s", listSp.GetName(), listSp.GetEntityID())
```
The entity ID itself is not sensitive, so exposing this should be okay.

mdwn replied:
Also, I've updated the error messages to be more consistent.

Commit:
Tim's and Edoardo's feedback.

Commit:
Merge remote-tracking branch 'origin/master' into mike.wilson/store-entity-id-in-saml-id-provider

Commit:
Fix cache test.

espadolini commented on lib/services/local/saml_idp_service_provider.go:
```diff
@@ -158,6 +190,45 @@ func (s *SAMLIdPServiceProviderService) DeleteAllSAMLIdPServiceProviders(ctx con
+			// Only check entity ID duplicates if we're looking at objects other than the one we're trying to validate.
+			// This ensures updates will work and that creates will return an already exists error.
+			if listSp.GetName() != sp.GetName() && listSp.GetEntityID() == sp.GetEntityID() {
+				return trace.BadParameter("%s %s has the same entity ID %s", types.KindSAMLIdPServiceProvider, listSp.GetName(), listSp.GetEntityID())
```
Perhaps?
```suggestion
				return trace.BadParameter("%s %q has the same entity ID %q", types.KindSAMLIdPServiceProvider, listSp.GetName(), listSp.GetEntityID())
```

espadolini reviewed (approved):
Add some comments around the `EntityID` field (in the proto or in the interface, maybe) to clarify that the entity ID is only checked to match the one in the entity description at upsert time, to avoid having to parse a giant xml blob whenever we unmarshal the resource for anything.

mdwn committed:
Update lib/services/local/saml_idp_service_provider.go

Co-authored-by: Edoardo Spadolini <edoardo.spadolini@goteleport.com>

Commit:
Expanding the comments on the EntityID field.

Commit:
Merge remote-tracking branch 'origin/master' into mike.wilson/store-entity-id-in-saml-id-provider

mdwn committed:
Merge branch 'master' into mike.wilson/store-entity-id-in-saml-id-provider
